File: scripts/bench/bench_phow_breakdown.py
#!/usr/bin/env python
from __future__ import annotations

import argparse
import os
import time
from typing import Dict, List, Tuple

import numpy as np

from densevlad.torii15 import Torii15Assets, load_torii15_vocab
from densevlad.torii15 import image as image_mod
from densevlad.torii15.densevlad import (
    _DSIFT_TRANSPOSE_PERM,
    _assignments_idx,
    _rootsift,
    _vl_vlad_hard,
)


def _mean(times: List[float]) -> float:
    return float(sum(times) / len(times)) if times else 0.0


def _timeit(fn) -> float:
    t0 = time.perf_counter()
    fn()
    return time.perf_counter() - t0


def _load_image(path: str, max_dim: int | None, apply_imdown: bool) -> np.ndarray:
    return image_mod.read_gray_im2single(path, max_dim=max_dim, apply_imdown=apply_imdown)


def bench_phow_breakdown(
    *,
    image_path: str,
    max_dim: int | None,
    apply_imdown: bool,
    reps: int,
    warmup: int,
    simd: bool,
    assign_method: str,
) -> Dict[str, float]:
    image_mod.set_simd_enabled(simd)
    os.environ["DVLAD_PHOW_WORKERS"] = "1"
    os.environ["DVLAD_PHOW_BACKEND"] = "serial"
    os.environ["DVLAD_ASSIGN_METHOD"] = assign_method

    vocab = load_torii15_vocab(Torii15Assets.default().vocab_mat_path())

    sizes = (4, 6, 8, 10)
    max_size = max(sizes)
    step = 2
    magnification = 6
    window_size = 1.5
    contrast_threshold = 0.005

    try:
        from cyvlfeat.sift import dsift
    except Exception as exc:  # pragma: no cover - bench-only
        raise SystemExit(f"cyvlfeat is required for benchmarks: {exc}")

    # Warmup
    for _ in range(max(0, warmup)):
        im = _load_image(image_path, max_dim, apply_imdown)
        for size in sizes:
            sigma = size / magnification
            ims = image_mod.vl_imsmooth_gaussian(im, sigma=sigma)
            ims_t = np.ascontiguousarray(ims.T)
            off = int(np.floor(1.0 + 1.5 * (max_size - size)))
            off0 = max(off - 1, 0)
            _ = dsift(
                ims_t,
                step=step,
                size=size,
                bounds=np.array(
                    [off0, off0, ims_t.shape[0] - 1, ims_t.shape[1] - 1], dtype=np.int32
                ),
                window_size=window_size,
                norm=True,
                fast=True,
                float_descriptors=False,
            )

    acc: Dict[str, float] = {}

    def add_time(key: str, dt: float) -> None:
        acc[key] = acc.get(key, 0.0) + dt

    for _ in range(reps):
        per: Dict[str, float] = {}

        def add_per(key: str, dt: float) -> None:
            per[key] = per.get(key, 0.0) + dt

        t0 = time.perf_counter()
        im = _load_image(image_path, max_dim, apply_imdown)
        add_per("preprocess", time.perf_counter() - t0)

        descs_all = []
        for size in sizes:
            sigma = size / magnification
            t0 = time.perf_counter()
            ims = image_mod.vl_imsmooth_gaussian(im, sigma=sigma)
            add_per(f"imsmooth_{size}", time.perf_counter() - t0)

            ims_t = np.ascontiguousarray(ims.T)
            off = int(np.floor(1.0 + 1.5 * (max_size - size)))
            off0 = max(off - 1, 0)

            t0 = time.perf_counter()
            frames, descs = dsift(
                ims_t,
                step=step,
                size=size,
                bounds=np.array(
                    [off0, off0, ims_t.shape[0] - 1, ims_t.shape[1] - 1], dtype=np.int32
                ),
                window_size=window_size,
                norm=True,
                fast=True,
                float_descriptors=False,
            )
            add_per(f"dsift_{size}", time.perf_counter() - t0)

            t0 = time.perf_counter()
            descs = descs[:, _DSIFT_TRANSPOSE_PERM]
            descs[frames[:, 2] < contrast_threshold] = 0
            add_per(f"post_{size}", time.perf_counter() - t0)
            descs_all.append(descs)

        t0 = time.perf_counter()
        descs = np.vstack(descs_all)
        add_per("phow_concat", time.perf_counter() - t0)

        t0 = time.perf_counter()
        descs_rs = _rootsift(descs)
        add_per("rootsift", time.perf_counter() - t0)

        t0 = time.perf_counter()
        idx = _assignments_idx(descs_rs, vocab.centers)
        add_per("assign", time.perf_counter() - t0)

        t0 = time.perf_counter()
        _ = _vl_vlad_hard(descs_rs, vocab.centers, idx)
        add_per("vlad", time.perf_counter() - t0)

        per["total"] = sum(
            per.get(k, 0.0) for k in (
                "preprocess",
                *[f"imsmooth_{s}" for s in sizes],
                *[f"dsift_{s}" for s in sizes],
                *[f"post_{s}" for s in sizes],
                "phow_concat",
                "rootsift",
                "assign",
                "vlad",
            )
        )

        for k, v in per.items():
            add_time(k, v)

    # average
    for k in list(acc.keys()):
        acc[k] /= reps
    return acc


def main() -> int:
    parser = argparse.ArgumentParser(description="Benchmark PHOW + VLAD breakdown (Python).")
    parser.add_argument("--image", default="", help="Override image path.")
    parser.add_argument("--max-dim", type=int, default=640, help="Max side before imdown.")
    parser.add_argument("--no-imdown", action="store_true", help="Disable vl_imdown.")
    parser.add_argument("--reps", type=int, default=3, help="Benchmark repetitions.")
    parser.add_argument("--warmup", type=int, default=1, help="Warmup repetitions.")
    parser.add_argument("--no-simd", action="store_true", help="Disable SIMD paths.")
    parser.add_argument(
        "--assign-method",
        default="kdtree",
        choices=["kdtree", "matmul", "hard", "kmeans"],
        help="Assignment method for VLAD.",
    )
    args = parser.parse_args()

    assets = Torii15Assets.default()
    image_path = args.image or assets.extract_member(
        "247code/data/example_gsv/L-NLvGeZ6JHX6JO8Xnf_BA_012_000.jpg"
    )

    res = bench_phow_breakdown(
        image_path=str(image_path),
        max_dim=None if args.max_dim <= 0 else args.max_dim,
        apply_imdown=not args.no_imdown,
        reps=max(1, args.reps),
        warmup=max(0, args.warmup),
        simd=not args.no_simd,
        assign_method=args.assign_method,
    )

    print("BENCH python")
    for key in sorted(res.keys()):
        print(f"BENCH {key} {res[key]:.6f}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/bench/bench_phow_breakdown.m
function bench_phow_breakdown(varargin)
%BENCH_PHOW_BREAKDOWN Detailed PHOW + VLAD timing breakdown (MATLAB).

opts = struct();
opts.reps = 3;
opts.warmup = 1;
opts.max_dim = 640;
opts.use_imdown = true;
opts.imfn = '';
opts.assign_method = 'kdtree';
opts = parse_opts(opts, varargin{:});

root_dir = fullfile(fileparts(mfilename('fullpath')), '..', '..', '247code');
root_dir = char(java.io.File(root_dir).getCanonicalPath());
orig_dir = pwd;
cd(root_dir);
run('at_setup');
cd(orig_dir);
try
    vl_setnumthreads(1);
catch
end

if isempty(opts.imfn)
    opts.imfn = fullfile(root_dir, 'data', 'example_gsv', 'L-NLvGeZ6JHX6JO8Xnf_BA_012_000.jpg');
end

dictfn = fullfile(root_dir, 'data', 'dnscnt_RDSIFT_K128.mat');
load(dictfn, 'CX');
CX = bsxfun(@rdivide, CX, sqrt(sum((CX.^2), 1)));
kdtree = vl_kdtreebuild(CX);

sizes = [4 6 8 10];
step = 2;
magnif = 6;
window_size = 1.5;
contrast_threshold = 0.005;
max_size = max(sizes);

% warmup
for w = 1:opts.warmup
    img = imread(opts.imfn);
    if size(img,3) > 1
        img = rgb2gray(img);
    end
    if opts.max_dim > 0
        [h, wimg] = size(img);
        scale = double(opts.max_dim) / double(max(h, wimg));
        if scale < 1
            img = imresize(img, scale, 'bilinear');
        end
    end
    if opts.use_imdown
        img = vl_imdown(img);
    end
    img_single = im2single(img);
    for s = sizes
        off = floor(1.0 + 3.0/2.0 * (max_size - s));
        sigma = s / magnif;
        ims = vl_imsmooth(img_single, sigma);
        [f_s, d_s] = vl_dsift(ims, 'Step', step, 'Size', s, 'Bounds', [off off +inf +inf], ...
            'Norm', 'Fast', 'WindowSize', window_size);
        contrast = f_s(3, :);
        d_s(:, contrast < contrast_threshold) = 0;
    end
end

acc_pre = 0;
acc_ims = zeros(size(sizes));
acc_dsift = zeros(size(sizes));
acc_post = zeros(size(sizes));
acc_concat = 0;
acc_rootsift = 0;
acc_assign = 0;
acc_assign_mat = 0;
acc_vlad = 0;

for r = 1:opts.reps
    t0 = tic;
    img = imread(opts.imfn);
    if size(img,3) > 1
        img = rgb2gray(img);
    end
    if opts.max_dim > 0
        [h, wimg] = size(img);
        scale = double(opts.max_dim) / double(max(h, wimg));
        if scale < 1
            img = imresize(img, scale, 'bilinear');
        end
    end
    if opts.use_imdown
        img = vl_imdown(img);
    end
    img_single = im2single(img);
    acc_pre = acc_pre + toc(t0);

    descs_all = cell(1, numel(sizes));
    for i = 1:numel(sizes)
        s = sizes(i);
        off = floor(1.0 + 3.0/2.0 * (max_size - s));
        sigma = s / magnif;
        t0 = tic;
        ims = vl_imsmooth(img_single, sigma);
        acc_ims(i) = acc_ims(i) + toc(t0);

        t0 = tic;
        [f_s, d_s] = vl_dsift(ims, 'Step', step, 'Size', s, 'Bounds', [off off +inf +inf], ...
            'Norm', 'Fast', 'WindowSize', window_size);
        acc_dsift(i) = acc_dsift(i) + toc(t0);

        t0 = tic;
        contrast = f_s(3, :);
        d_s(:, contrast < contrast_threshold) = 0;
        acc_post(i) = acc_post(i) + toc(t0);
        descs_all{i} = d_s;
    end

    t0 = tic;
    desc = cat(2, descs_all{:});
    acc_concat = acc_concat + toc(t0);

    t0 = tic;
    desc_rs = relja_rootsift(single(desc));
    acc_rootsift = acc_rootsift + toc(t0);

    if strcmpi(opts.assign_method, 'kdtree')
        t0 = tic;
        nn = vl_kdtreequery(kdtree, CX, desc_rs);
        acc_assign = acc_assign + toc(t0);

        t0 = tic;
        assigns = zeros(size(CX, 2), size(desc_rs, 2), 'single');
        assigns(sub2ind(size(assigns), double(nn), 1:length(nn))) = 1;
        acc_assign_mat = acc_assign_mat + toc(t0);
    else
        error('Only assign_method=kdtree is supported in MATLAB benchmark.');
    end

    t0 = tic;
    vlad = vl_vlad(desc_rs, CX, assigns, 'NormalizeComponents');
    acc_vlad = acc_vlad + toc(t0);
end

reps = double(opts.reps);

fprintf(1, 'BENCH matlab\n');
fprintf(1, 'BENCH preprocess %.6f\n', acc_pre / reps);
for i = 1:numel(sizes)
    fprintf(1, 'BENCH imsmooth_%d %.6f\n', sizes(i), acc_ims(i) / reps);
end
for i = 1:numel(sizes)
    fprintf(1, 'BENCH dsift_%d %.6f\n', sizes(i), acc_dsift(i) / reps);
end
for i = 1:numel(sizes)
    fprintf(1, 'BENCH post_%d %.6f\n', sizes(i), acc_post(i) / reps);
end
fprintf(1, 'BENCH phow_concat %.6f\n', acc_concat / reps);
fprintf(1, 'BENCH rootsift %.6f\n', acc_rootsift / reps);
fprintf(1, 'BENCH assign %.6f\n', acc_assign / reps);
fprintf(1, 'BENCH assign_mat %.6f\n', acc_assign_mat / reps);
fprintf(1, 'BENCH vlad %.6f\n', acc_vlad / reps);

end

function opts = parse_opts(opts, varargin)
if isempty(varargin)
    return
end
if mod(numel(varargin), 2) ~= 0
    error('Options must be name/value pairs.')
end
for i = 1:2:numel(varargin)
    key = varargin{i};
    val = varargin{i+1};
    if ~ischar(key)
        error('Option keys must be strings.')
    end
    key = lower(key);
    switch key
        case 'reps'
            opts.reps = val;
        case 'warmup'
            opts.warmup = val;
        case 'max_dim'
            opts.max_dim = val;
        case 'use_imdown'
            opts.use_imdown = logical(val);
        case 'imfn'
            opts.imfn = val;
        case 'assign_method'
            opts.assign_method = val;
        otherwise
            error('Unknown option: %s', key);
    end
end
end
